crates/sui-types/src/messages.rs

```diff
@@ -1120,8 +1232,32 @@ impl Message for SenderSignedData {
         if self.intent_message.value.kind.is_system_tx() {
             return Ok(());
         }
-        self.tx_signature
-            .verify_secure(&self.intent_message, self.intent_message.value.sender)
+
+        // Verify signatures. Steps are ordered in asc complexity order to minimize abuse.
+        let signers = self.intent_message.value.signers();
+        // Signature number needs to match
+        fp_ensure!(
+            self.tx_signatures.len() == signers.len(),
+            SuiError::SignerSignatureNumberMismatch {
+                actual: self.tx_signatures.len(),
+                expected: signers.len()
+            }
+        );
+        // All required signers need to be sign.
+        let present_sigs = self.get_signer_sig_mapping()?;
+        for s in signers {
+            if !present_sigs.contains_key(&s) {
+                return Err(SuiError::SignerSignatureAbsent {
+                    signer: s.to_string(),
+                });
+            }
+        }
+
+        // Verify all present signatures.
+        for (signer, signature) in present_sigs {
+            signature.verify_secure(&self.intent_message, signer)?;
+        }
```

Comment:
So for a sponsored txn there needs to be a little bit of back and forth between the sponsor and the user to collaboratively construct the txn that both parties end up signing. As in the sponsor doesn't only sign the gas object, they're signing the whole txn...correct?

Reply:
correct!